### PR TITLE
add ddpg

### PR DESCRIPTION
**Pull request by cjcchen**
add ddpg implementation to master branch

### CODE REVIEW

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+                    next_state, reward, terminal, info = self._env.step(
+                        action * self._max_action)
+                    self._replay_buffer.add_data(state, action,
+                                                 reward * self._reward_scale,
+                                                 terminal, next_state)
+                    state = next_state
+                    total_reward += reward
+                    if terminal:
+                        self._report_total_reward(total_reward, episode_step)
+                        print("epoch %d, total reward %lf\n" % (episode_step,
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: Use rllab's logger for this (`logger.record_tabular('key', value)`) and log more performance measures than reward, e.g.:
- actor loss
- critic loss
- max and average predicted Q value
- etc. (maybe you can think of more things that could help us understand the performance)

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        self._critic_net = CriticNet(
+            self._session,
+            gamma=gamma,
+            lr=critic_lr,
+            weight_decay=critic_l2_weight_decay)
+        self._target_critic = copy(self._critic_net)
+        self._target_critic.name = 'target_critic'
+
+        #replay buffer
+        self._replay_buffer = ReplayBuffer(1e6)
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: The size of the replay buffer should be user-definable

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+import numpy as np
+import os
+
+
+class DDPG(RLAlgorithm):
+    def __init__(self,
+                 env,
+                 gamma=0.99,
+                 tau=0.001,
+                 observation_range=(-5, 5),
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: `observation_range` isn't used anywhere (and shouldn't be needed)

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            lr=critic_lr,
+            weight_decay=critic_l2_weight_decay)
+        self._target_critic = copy(self._critic_net)
+        self._target_critic.name = 'target_critic'
+
+        #replay buffer
+        self._replay_buffer = ReplayBuffer(1e6)
+
+        if log_dir:
+            self._summary_writer = tf.summary.FileWriter(
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: Use rllab's logger

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+                 observation_range=(-5, 5),
+                 action_range=(-1, 1),
+                 actor_lr=1e-4,
+                 critic_lr=1e-3,
+                 reward_scale=1,
+                 batch_size=64,
+                 critic_l2_weight_decay=0.01,
+                 action_noise=None,
+                 plot=False,
+                 check_point_dir=None,
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: Use `logger.save_itr_params` which takes care of checkpoint folders etc.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            self._lr, beta1=0.9, beta2=0.999, epsilon=1e-08)
+
+        grads = tf.gradients(self._loss, self.trainable_vars)
+        grads_and_vars = list(zip(grads, self.trainable_vars))
+        self._train_op = self._optimizer.apply_gradients(grads_and_vars)
+
+        self._action_grads = tf.gradients(self._action_loss,
+                                          self._action_predict)
+
+    def _build_net(self, state, action, reuse=False):
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py:176]: add support for user-defined number and sizes of hidden layers, e.g. `hidden_layers=[64, 64]`

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            self._optimizer = tf.train.AdamOptimizer(
+                self._lr, beta1=0.9, beta2=0.999, epsilon=1e-08)
+
+    def set_grad(self, grad):
+        grads = tf.gradients(
+            ys=self._action, xs=self.trainable_vars, grad_ys=grad)
+        self._train_op = self._optimizer.apply_gradients(
+            zip(grads, self.trainable_vars))
+
+    def _build_net(self, state):
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py:68]: add support for user-defined number and sizes of hidden layers, e.g. `hidden_layers=[64, 64]`

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+                        total_reward = 0
+                        state = self._env.reset()
+                        if self._action_noise:
+                            self._action_noise.reset()
+
+                for train in range(train_steps):
+                    self._train_net()
+            self.save_session(episode_step)
+
+    def load_session(self):
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: We don't need this code if everything is serializable. Make sure you can run your implementation with `run_experiment_lite`.

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+class DDPG(RLAlgorithm):
+    def __init__(self,
+                 env,
+                 gamma=0.99,
+                 tau=0.001,
+                 observation_range=(-5, 5),
+                 action_range=(-1, 1),
+                 actor_lr=1e-4,
+                 critic_lr=1e-3,
+                 reward_scale=1,
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: `reward_scale` should be called `discount` (in accordance with the other implementations). A value of 1 seems strange, are you sure it isn't <1, e.g. 0.99?

File: sandbox/rocky/tf/algos/ddpg/ddpg.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+import os
+
+
+class DDPG(RLAlgorithm):
+    def __init__(self,
+                 env,
+                 gamma=0.99,
+                 tau=0.001,
+                 observation_range=(-5, 5),
+                 action_range=(-1, 1),
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/ddpg/ddpg.py]: Assume that the algorithm always gets a normalized environment, where `action_range` is always in [-1, 1]. We need to implement normalization for gym Envs anyways as described in #64.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+        vars = self.trainable_vars
+        target_vars = target_net.trainable_vars
+
+        soft_updates = []
+        vars = vars[0:len(target_vars)]
+        for var, target_var in zip(vars, target_vars):
+            soft_updates.append(
+                tf.assign(target_var, (1. - tau) * target_var + tau * var))
+        assert len(soft_updates) == len(vars)
+        self._update_paras = tf.group(*soft_updates)
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py]: Make sure you set a meaningful name on every TensorFlow operation

**Review by eric-heiden — CHANGES_REQUESTED** [review]
Make sure you set a name on every TensorFlow placeholder and operation so the graph is readable. I'm not sure if this implementation is serializable right now, so please ensure you can run DDPG with `run_experiment_lite` (add a separate launcher for this). See more comments in the code.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
+import tensorflow as tf
+import tensorflow.contrib as tc
+import numpy as np
+
+
+def _fc(x,
+        output_dim,
+        weight_initializer=tf.contrib.layers.xavier_initializer(uniform=False),
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py:8]: initialize weights based on seed from [get_seed()](https://github.com/ryanjulian/rllab/blob/integration/rllab/misc/ext.py#L209)

File: examples/ddpg_gym_tf_pendulum.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+from sandbox.rocky.tf.algos.ddpg.ddpg import DDPG
+from sandbox.rocky.tf.algos.ddpg.noise import OrnsteinUhlenbeckActionNoise
+
+import gym
+import numpy as np
+import tensorflow as tf
+
+RANDOM_SEED = 1234
```
Comment (eric-heiden) [inline examples/ddpg_gym_tf_pendulum.py]: Use rllab's [get_seed()](https://github.com/ryanjulian/rllab/blob/integration/rllab/misc/ext.py#L209)

File: examples/ddpg_gym_tf_pendulum.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+np.random.seed(RANDOM_SEED)
+tf.set_random_seed(RANDOM_SEED)
```
Comment (eric-heiden) [inline examples/ddpg_gym_tf_pendulum.py]: remove these lines, call [set_seed(seed)](https://github.com/ryanjulian/rllab/blob/integration/rllab/misc/ext.py#L188) instead

File: examples/ddpg_gym_tf_pendulum.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+env = gym.make('Pendulum-v0')
+env.seed(RANDOM_SEED)
```
Comment (eric-heiden) [inline examples/ddpg_gym_tf_pendulum.py]: use [get_seed()](https://github.com/ryanjulian/rllab/blob/integration/rllab/misc/ext.py#L209)

File: examples/ddpg_gym_tf_pendulum.py [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+action_dim = env.action_space.shape[-1]
+action_noise = OrnsteinUhlenbeckActionNoise(
```
Comment (eric-heiden) [inline examples/ddpg_gym_tf_pendulum.py]: the noise should also be based on a seed

File: sandbox/rocky/tf/algos/ddpg/ddpg.py (hunk 5 above)
Comment (cjcchen) [reply sandbox/rocky/tf/algos/ddpg/ddpg.py]: I am not sure what this function is used for yet. The checkpoint folder is used for tensorflow, it seems it does not need to save other parameters but the graph.

File: sandbox/rocky/tf/algos/ddpg/ddpg.py (hunk 9 above)
Comment (cjcchen) [reply sandbox/rocky/tf/algos/ddpg/ddpg.py]: In the paper and baseline, they all set this value to be 1.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py (hunk 12 above)
Comment (cjcchen) [reply sandbox/rocky/tf/algos/network/actor_critic_net.py:8]: It seems that this weight does not use seed to be initialized.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 17]
```diff
@@ -64,12 +65,12 @@ def _build_net(self, state):
         with tf.variable_scope(self.name) as scope:
 
             with tf.variable_scope('fc1'):
-                fc1_out = _fc(state, 64)
+                fc1_out = _fc(state, self._hidden_layers[0])
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py]: There could be more than 2 hidden layers, so you should create `fc` layers dynamically based on the array size.

File: sandbox/rocky/tf/algos/network/actor_critic_net.py [hunk 18]
```diff
@@ -166,13 +168,13 @@ def _build_net(self, state, action, reuse=False):
                 scope.reuse_variables()
 
             with tf.variable_scope('fc1'):
-                fc1_out = _fc(state, 64)
+                fc1_out = _fc(state, self._hidden_layers[0])
```
Comment (eric-heiden) [inline sandbox/rocky/tf/algos/network/actor_critic_net.py]: There could be more than 2 hidden layers, so you should create `fc` layers dynamically based on the array size.

File: examples/ddpg_gym_tf_pendulum.py [hunk 19]
```diff
@@ -1,17 +1,18 @@
...
-tf.set_random_seed(RANDOM_SEED)
+ext.set_seed(RANDOM_SEED)
 
 env = gym.make('Pendulum-v0')
-env.seed(RANDOM_SEED)
+env.seed(ext.get_seed())
+
 
 action_dim = env.action_space.shape[-1]
 action_noise = OrnsteinUhlenbeckActionNoise(
```
Comment (eric-heiden) [inline examples/ddpg_gym_tf_pendulum.py]: Please add another launcher script which runs your code via `run_experiment_lite`. It should work like this:
```python
def run_task(*_):
    # initialize actor, critic, noise, env, etc.
    env = ...

    algo = DDPG(env=env, ...)
    algo.train()

run_experiment_lite(
    run_task,
    n_parallel=20,
    plot=False,
)
```
We need to make sure your DDPG implementation is serializable.

**Review by eric-heiden — CHANGES_REQUESTED** [review]
